Simplenote/src/main/java/com/automattic/simplenote/PreferencesFragment.java

```diff
@@ -456,6 +481,101 @@ public int compare(String text1, String text2) {
+        } catch (Exception e) {
+            Toast.makeText(requireContext(), getString(R.string.import_message_failure), Toast.LENGTH_SHORT).show();
+        }
+    }
+
+    private void importPlaintextFile(String exportContents) {
+        Simplenote currentApp = (Simplenote) requireActivity().getApplication();
+        Bucket<Note> noteBucket = currentApp.getNotesBucket();
+        Note note = noteBucket.newObject();
+        note.setContent(exportContents);
+        note.setCreationDate(Calendar.getInstance());
+        note.setModificationDate(note.getCreationDate());
+        note.save();
+    }
+
+    private void importJsonExport(String exportContents) throws Exception {
+        JSONObject jsonData = new JSONObject(exportContents);
+        JSONArray activeNotes = jsonData.getJSONArray("activeNotes");
+        JSONArray trashedNotes = jsonData.getJSONArray("trashedNotes");
+        List<String> trashedIds = new ArrayList<>();
+        JSONArray notesArray = new JSONArray();
+        for (int i = 0; i < activeNotes.length(); i++) {
+            notesArray.put(activeNotes.getJSONObject(i));
+        }
+        for (int j = 0; j < trashedNotes.length(); j++) {
+            notesArray.put(trashedNotes.getJSONObject(j));
+            trashedIds.add(trashedNotes.getJSONObject(j).getString("id"));
+        }
+        for (int k = 0; k < notesArray.length(); k++) {
+            JSONObject noteJson = notesArray.getJSONObject(k);
+            noteFromJson(noteJson, trashedIds).save();
+        }
+    }
+
+    private Note noteFromJson(JSONObject noteJson, List<String> trashedIds) throws Exception {
+        Simplenote currentApp = (Simplenote) requireActivity().getApplication();
+        Bucket<Note> noteBucket = currentApp.getNotesBucket();
+        Note note = noteBucket.newObject();
+        note.setContent(noteJson.has("content") ? noteJson.getString("content") : "");
+        note.setCreationDate(noteJson.has("creationDate") ? DateTimeUtils.getDateCalendar(noteJson.getString("creationDate")) : Calendar.getInstance());
+        note.setModificationDate(noteJson.has("lastModified") ? DateTimeUtils.getDateCalendar(noteJson.getString("lastModified")) : Calendar.getInstance());
+        note.setTags(noteJson.has("tags") ? jsonToList(noteJson.getJSONArray("tags")) : new ArrayList<String>());
+        note.setPinned(noteJson.has("pinned") && noteJson.getBoolean("pinned"));
+        note.setMarkdownEnabled(noteJson.has("markdown") && noteJson.getBoolean("markdown"));
+        note.setDeleted(noteJson.has("id") && trashedIds.contains(noteJson.getString("id")));
+        return note;
+    }
+
+    private List<String> jsonToList(JSONArray json) {
+        List<String> list = new ArrayList<>();
+        for (int i = 0; i < json.length(); i++) {
+            try {
+                list.add(json.getString(i));
+            } catch (JSONException e) {
+                e.printStackTrace();
```

Comment:
we'll want to match our error logging with the rest of the app vs. dumping a stack trace.

in addition, it looks like the `Note` class stores its list of tags as a `JSONArray` so we could expose a new method for `setTags(JSONArray tags)` and skip the encode/decode loop.

on this point, are we able to directly instantiate a note with the existing Note class, without conversion? I doubt it works 100% but the constructor can take a JSON object. it may be the case that we can perform minimal transformation and/or slightly modify `Note.java` so we can go straight from the imported JSON object to a note.

```java
new Note(activeNotes.getObject(i)).save();
```

Reply:
I changed tags adding, but I doubt about the note. In theory, we can use the following code:

`Simplenote currentApp = (Simplenote) activity.getApplication();`
`Bucket<Note> noteBucket = currentApp.getNotesBucket();`
`Note note = new Note("note - " +  UUID.randomUUID(), activeNotes.getJSONObject(i));`
`noteBucket.add(note);`
`note.save();`

It looks much better, but I doubt this is a good idea, since I am not sure how correctly the Ыimperium alogyrythms will work in this case. It seems to me that it is preferable to use `noteBucket.newObject();`

Reply:
I'll leave this as it was for now, and then we'll decide!

Reply:
thanks. I might need to consult with a few people on this. please pardon the delay, but expect a few days gap.

Reply:
Sure, no problem!